raven/src/main/java/dk/itu/raven/Raven.java

```diff
@@ -47,65 +41,64 @@ public static void main(String[] args) throws IOException {
-        long endBuildNano = System.nanoTime();
-        Logger.log("Build time: " + (endBuildNano - startBuildNano) / 1000000 + "ms", LogLevel.INFO);
-        Logger.log("Done Building Raster", LogLevel.INFO);
-        Logger.log(k2Raster.tree.size(), LogLevel.DEBUG);
+        ImageMetadata metadata = rasterReader.getImageMetadata();
 
-        int w = data.second.getWidth();
-        int h = data.second.getHeight();
-        int[] sampleSize = data.second.getSampleSize();
-        int bitsUsed = data.second.getBitsUsed();
-        data.second = null;
+        int totalBits = 0;
+        for (int bits : metadata.getBitsPerSample()) {
+            totalBits += bits;
+        }
 
-        RasterFilterFunction function;
+        IRasterFilterFunction function = JoinFilterFunctions.acceptAll();
 
         if (jct.ranges.size() == 2) {
-            if (sampleSize.length > 1) {
+            if (metadata.getBitsPerSample().length > 1) {
                 Logger.log("WARNING: only one range was given, but more than one raster sample exists ("
-                        + sampleSize.length + ")", LogLevel.WARNING);
+                        + metadata.getBitsPerSample().length + ")", LogLevel.WARNING);
             }
             long lo = jct.ranges.get(0);
             long hi = jct.ranges.get(1);
             function = JoinFilterFunctions.rangeFilter(lo, hi);
-        } else if (jct.ranges.size() == sampleSize.length * 2) {
+        } else if (jct.ranges.size() == metadata.getBitsPerSample().length * 2) {
             Logger.log("using multiSampleRangeFilter", LogLevel.DEBUG);
-            function = JoinFilterFunctions.multiSampleRangeFilter(jct.ranges, sampleSize, bitsUsed);
+            function = JoinFilterFunctions.multiSampleRangeFilter(jct.ranges, metadata.getBitsPerSample(),
+                    totalBits);
         } else if (jct.ranges.size() == 0) {
             function = JoinFilterFunctions.rangeFilter(Integer.MIN_VALUE, Integer.MAX_VALUE);
         } else {
             throw new IllegalArgumentException(
                     "The number of provided search ranges does not match the number of raster samples");
         }
 
-        // create a R* tree with
-        RTree<String, Geometry> rtree = api.generateRTree(data.first);
-        Logger.log("Done Building rtree", LogLevel.INFO);
-
-        // construct and compute the join
-        RavenJoin join = new RavenJoin(k2Raster, rtree);
         long startJoinNano = System.nanoTime();
-        List<Pair<Geometry, Collection<PixelRange>>> result = join.join(function);
+        AbstractRavenJoin join;
+        if (jct.streamed) {
+            join = api.getStreamedJoin(jct.inputRaster, jct.inputVector, jct.tileSize, jct.tileSize, jct.parallel);
+        } else {
+            join = api.getJoin(jct.inputRaster, jct.inputVector);
+        }
+        AbstractJoinResult result = join.join(function);
+
+        // result.count();
+        result = result.asMemoryAllocatedResult();
```

Comment:
We only want a memory allocatedResult when drawing, otherwise we risk crashing on large datasets.